drivers/ipm/ipm_esp32.c

```diff
@@ -0,0 +1,280 @@
+
+		case CONFIG_ESP32_IPM_CHANNEL_WIFI:
+			if(dev_data->cb[CONFIG_ESP32_IPM_CHANNEL_WIFI]) {
+				dev_data->cb[CONFIG_ESP32_IPM_CHANNEL_WIFI](dev,
+							NULL,
+							CONFIG_ESP32_IPM_CHANNEL_WIFI,
+							dev_data->shm->wifi_shm);
+			}
+			break;
+
+		case CONFIG_ESP32_IPM_CHANNEL_BT:
+			if(dev_data->cb[CONFIG_ESP32_IPM_CHANNEL_BT]) {
+				dev_data->cb[CONFIG_ESP32_IPM_CHANNEL_BT](dev,
+							NULL, 
+							CONFIG_ESP32_IPM_CHANNEL_BT, 
+							dev_data->shm->bt_shm);
+			}
+			break;
+
+		case CONFIG_ESP32_IPM_CHANNEL_USER_0:
+			if(dev_data->cb[CONFIG_ESP32_IPM_CHANNEL_USER_0]) {
+				dev_data->cb[CONFIG_ESP32_IPM_CHANNEL_USER_0](dev,
+							NULL, 
+							CONFIG_ESP32_IPM_CHANNEL_USER_0, 
+							dev_data->shm->user_0_shm);
+			}
+			break;
+
+		case CONFIG_ESP32_IPM_CHANNEL_USER_1:
+			if(dev_data->cb[CONFIG_ESP32_IPM_CHANNEL_USER_1]) {
+				dev_data->cb[CONFIG_ESP32_IPM_CHANNEL_USER_1](dev,
+							NULL, 
+							CONFIG_ESP32_IPM_CHANNEL_USER_1, 
+							dev_data->shm->user_1_shm);
+			}
+			break;
+	}
+
+	/* unlock the shared memory */
+	atomic_set(&dev_data->control->lock, ESP32_IPM_LOCK_FREE_VAL);	
+}
+
+static int esp32_ipm_send(const struct device *dev, int wait, uint32_t id,
+			 const void *data, int size)
+{
+	size_t channel_size = esp32_ipm_channel_data_size(id);
+	struct esp32_ipm_data *dev_data = (struct esp32_ipm_data *)dev->data;
+
+	if(data == NULL) {
+		LOG_ERR("Invalid data source");
+		return -EINVAL;
+	}
+
+	if(!channel_size) {
+		LOG_ERR("Invalid IPM channel");
+		return -EINVAL;
+	}
+
+	if(channel_size < size) {
+		LOG_ERR("Not enough memory in IPM channel");
+		return -ENOMEM;
+	}
+
+	/* try to lock the shared memory */
+	while(!atomic_cas(&dev_data->control->lock, ESP32_IPM_LOCK_FREE_VAL, dev_data->this_core_id)) {
+		
+		k_busy_wait(1);
+
+		if((wait != -1) && (wait > 0)) {
+			/* lock could not be held this time, return */
+			wait--;
+			if(wait == 0)  {
+				return -ETIMEDOUT;	
+			}
+		}
+	}
+
+	/* We own the shared memory, so write data to it: */
+	esp32_ipm_copy_data_to_memory(dev_data->shm, id, data, size);
+	dev_data->control->dest_cpu_msg_id[dev_data->other_core_id] = id;
+
+	/* data copied, set the id and, generate interrupt in the remote core */
+	if (dev_data->this_core_id == 0) {
+		LOG_DBG("Generating interrupt on remote CPU 1 from CPU 0");
+		DPORT_WRITE_PERI_REG(DPORT_CPU_INTR_FROM_CPU_1_REG, DPORT_CPU_INTR_FROM_CPU_1);	
+	} else {
+		LOG_DBG("Generating interrupt on remote CPU 0 from CPU 1");
+		DPORT_WRITE_PERI_REG(DPORT_CPU_INTR_FROM_CPU_0_REG, DPORT_CPU_INTR_FROM_CPU_0);
+	}
+
+	/* release the lock */
+	atomic_set(&dev_data->control->lock, ESP32_IPM_LOCK_FREE_VAL);
+
+	return 0;
+}
+
+static void esp32_ipm_register_callback(const struct device *dev,
+				       ipm_callback_t cb,
+				       void *user_data)
+{
+	struct esp32_ipm_data *data = (struct esp32_ipm_data *)dev->data;
+	uint32_t id; 
+
+	if(user_data == NULL) {
+		id = 0;
+	} else {
+		id = (uint32_t )user_data;
+	}
```

Comment:
Isn't this exactly the use case for MBOX instead of IPM? You are registering different callbacks on different channels.

Reply:
Oh, it seems this is a leftovers from the development version, the current IPC implementation has a single callback and a single channel, the id is just a placeholder to pass IPC frame identification to the upper layer which is responsible to decode.

Reply:
Oh, it seems this one is a previous commit, the actual implementation is not like that,